psi4/driver/procrouting/scf_proc/subclass_methods.py

```diff
@@ -119,3 +122,107 @@ def _ROHF_diis(self, Dnorm):
 core.RHF.diis = _RHF_diis
 core.UHF.diis = core.CUHF.diis = _UHF_diis
 core.ROHF.diis = _ROHF_diis
+
+def _UHF_stability_analysis(self):
+    # => Validate options <=
+    # TODO: Stability analysis is supported for any functional UKS functional where its one-
+    # and two-body Hamiltonian matrix-vector products are implemented. This is true for LDA
+    # at least, but probably not other functionals. This restriction exists for now because
+    # we've always had it, but we should lift it as much as we can and implement more matrix-
+    # vector products so we can lift it further.
+    # TODO: It should be up to the SolverEngine to validate whether it can do Hx products for the input wfn.
+    if self.functional().is_gga() or self.functional().is_meta():
+        raise ValidationError("Stability analysis not yet supported for non-LDA functionals.")
+
+    # => Prep options for eigenvector solver <=
+    if not self.options().has_changed("SOLVER_ROOTS_PER_IRREP"):
+        roots = [self.options().get_int("SOLVER_N_ROOT")] * self.nirrep()
+    else:
+        roots = self.options().get_int_vector("SOLVER_ROOTS_PER_IRREP")
```

Comment:
As we've discussed some privately, Francesco does not like options contamination between computations. I decided to get the options object from the `wfn` rather than the global with possible futureproofing in mind.

I would hope the two versions would be equivalent under current code, and I can change to your version if you'd like. Your version has the advantage of making my non-const `has_changed` implementation no longer necessary. I can use the functions you pointed me to, that check the global object.

Reply:
Yes, when you were asking about equivalence of P::e.options and wfn.options, I was thinking c-side within a certain module. I'd prefer the suggested version if you're ok with it.

I think we can do the "options-passing" `energy(..., keywords={...})` that FAE wants. But I don't think it'll be done as cleanly as mol-passing or wfn-passing where the object never resides in globals. Instead I think it'll be a save-and-restore like https://github.com/psi4/psi4/blob/master/psi4/driver/p4util/procutil.py#L372 used presently for calling qcng in distributed driver.

Reply:
I can do that, but this raises a beyond-the-scope-of-this-PR design question: why does the wavefunction carry an options object if options are always supposed to be read from globals?